### PR TITLE
Add support for retry of failing downstream calls

### PR DESCRIPTION
**Pull request by Umang01-hash**
## Pull Request Template

**Description:**

-   Resolves issue #709 

-   Example Usage:
```
a.AddHTTPService("cat-facts", "http://localhost:9000/",
		&service.HealthConfig{
			HealthEndpoint: "breeds",
		},
		&service.RetryConfig{MaxRetries: 5},
```

**Checklist:**

-   [x] I have formatted my code using  `goimport`  and  `golangci-lint`.
-   [x] All new code is covered by unit tests.
-   [x] This PR does not decrease the overall code coverage.
-   [x] I have reviewed the code comments and documentation for clarity.

**Thank you for your contribution!**

### CODE REVIEW

File: pkg/gofr/service/retry_test.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+}
+
+func TestRetryProvider_Get(t *testing.T) {
+	mockHTTP := &mockHTTP{}
+	retryConfig := &RetryConfig{MaxRetries: 3}
+	retryHTTP := retryConfig.AddOption(mockHTTP)
+
+	// Make the GET request
+	resp, err := retryHTTP.Get(context.Background(), "/test", nil)
+	assert.Nil(t, err)
```
Comment (ccoVeille) [inline pkg/gofr/service/retry_test.go:80]: Here and everywhere

```suggestion
	assert.NoError(t, err)
```

You can add `testifylint` to golangci-lint linter

**Review by ccoVeille — APPROVED** [review]

**Review by aryanmehrotra — APPROVED** [review]

File: pkg/gofr/service/retry.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+func (rp *retryProvider) DeleteWithHeaders(ctx context.Context, path string, body []byte, headers map[string]string) (
+	*http.Response, error) {
+	return rp.doWithRetry(func() (*http.Response, error) {
+		return rp.HTTP.DeleteWithHeaders(ctx, path, body, headers)
+	})
+}
+
+func (rp *retryProvider) doWithRetry(reqFunc func() (*http.Response, error)) (*http.Response, error) {
+	var (
+		resp *http.Response
```
Comment (ccoVeille) [inline pkg/gofr/service/retry.go:97]: Why not simply defining them in the function signature ? As named returned values, I mean.

File: pkg/gofr/service/retry.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+		err  error
+	)
+
+	for i := 0; i < rp.maxRetries; i++ {
+		resp, err = reqFunc()
+		if err == nil && resp.StatusCode != 500 {
```
Comment (ccoVeille) [inline pkg/gofr/service/retry.go:103]:
```suggestion
		if err == nil && resp.StatusCode != http.StatusInternalError {
```

File: pkg/gofr/service/retry.go (hunk 2 above)
Comment (Umang01-hash) [reply pkg/gofr/service/retry.go:97]: Didn't got you?..

File: pkg/gofr/service/retry.go (hunk 3 above)
Comment (Umang01-hash) [reply pkg/gofr/service/retry.go:103]: Sure this will be rectified. Thankyou

Comment (ccoVeille) [reply pkg/gofr/service/retry.go:103]: usestdlibvars linter can be useful, this one is available in golangci-lint

https://golangci-lint.run/usage/linters/#usestdlibvars

File: pkg/gofr/service/retry.go (hunk 2 above)
Comment (ccoVeille) [reply pkg/gofr/service/retry.go:97]:
```diff
-func (rp *retryProvider) doWithRetry(reqFunc func() (*http.Response, error)) (*http.Response, error) {
-
-       var ( 
-                 resp *http.Response 
-                 err  error 
-         )
-
+func (rp *retryProvider) doWithRetry(reqFunc func() (*http.Response, error)) (resp *http.Response, err error) {
+
```

**Pull request merged**